fluence-faas/src/faas.rs

```diff
@@ -36,60 +49,86 @@ pub struct FluenceFaaS {
 }
 
 impl FluenceFaaS {
-    pub fn new<P: Into<PathBuf>>(
-        core_modules_dir: P,
-        config_file_path: P,
-    ) -> Result<Self, FaaSError> {
+    /// Creates FaaS from config on filesystem.
+    pub fn new<P: Into<PathBuf>>(config_file_path: P) -> Result<Self> {
+        let config = crate::misc::load_config(config_file_path.into())?;
+        Self::with_raw_config(config)
+    }
+
+    /// Creates FaaS from config deserialized from TOML.
+    pub fn with_raw_config(config: RawCoreModulesConfig) -> Result<Self> {
+        let config = crate::misc::from_raw_config(config)?;
+        let modules = config
+            .core_modules_dir
+            .as_ref()
+            .map_or(Ok(vec![]), |dir| Self::load_modules(dir))?;
+        Self::with_modules(modules, config)
+    }
+
+    /// Creates FaaS with given modules.
+    pub fn with_modules<I, C>(modules: I, config: C) -> Result<Self>
+    where
+        I: IntoIterator<Item = (String, Vec<u8>)>,
+        C: TryInto<CoreModulesConfig>,
+        FaaSError: From<C::Error>,
+    {
         let mut fce = FCE::new();
         let mut module_names = Vec::new();
-        let mut core_modules_config = crate::misc::parse_config_from_file(config_file_path.into())?;
-
-        for entry in fs::read_dir(core_modules_dir.into())? {
-            let path = entry?.path();
-            if path.is_dir() {
-                // just skip directories
-                continue;
-            }
-
-            let module_name = path.file_name().unwrap();
-            let module_name = module_name
-                .to_os_string()
-                .into_string()
-                .map_err(|e| FaaSError::IOError(format!("failed to read from {:?} file", e)))?;
-
-            let module_bytes = fs::read(path.clone())?;
-
-            let core_module_config = crate::misc::make_fce_config(
-                core_modules_config.modules_config.remove(&module_name),
-            )?;
-            fce.load_module(module_name.clone(), &module_bytes, core_module_config)?;
-            module_names.push(module_name);
+        let mut config = config.try_into()?;
+
+        for (name, bytes) in modules {
+            let module_config = crate::misc::make_fce_config(config.modules_config.remove(&name))?;
+            fce.load_module(name.clone(), &bytes, module_config)?;
+            module_names.push(name);
         }
 
-        let rpc_module_config =
-            crate::misc::make_fce_config(core_modules_config.rpc_module_config)?;
+        let faas_code_config = make_fce_config(config.rpc_module_config)?;
 
         Ok(Self {
             fce,
             module_names,
-            faas_code_config: rpc_module_config,
+            faas_code_config,
         })
     }
 
+    /// Creates FaaS from prepared config.
+    fn load_modules(core_modules_dir: &str) -> Result<Vec<(String, Vec<u8>)>> {
```

Comment:
Algoritmically, there are two iterations. At first, for collecting results of dir reading, secondly, to map these results to vec. Moreover, there are some excess objects will be creating during the first iteration and we have to do the third iteration in caller if this function.

So, I sugget to use the "imperative" approach with mutable Vec to collecting results (as it was before in the `new` method). The code will be clear and all "imperative" logic will be isolated in this separate function. Other option is [this](https://lliwynd.blogspot.com/2019/01/iterating-over-results-in-rust.html), but code will be less readable in my opinion.

Reply:
Nice catch! I've used `try_fold`, it looks much better now.